### PR TITLE
Payments

### PR DESCRIPTION
**Pull request by ptran068**
Review

### CODE REVIEW

File: src/payments/admin.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+from django.contrib import admin
+
+# Register your models here.
```
Comment (datnguyen0126) [inline src/payments/admin.py:3]: remove this

File: src/payments/serializers.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+class PaymentSerializer(serializers.ModelSerializer):
+    tour = TourSerializer(many=False)
+
+    class Meta:
+        model = Payments
+        fields = '__all__'
+
+class CustomerSerializer(serializers.Serializer):
+    description = serializers.CharField(max_length = 255)
```
Comment (datnguyen0126) [inline src/payments/serializers.py:23]: why you don't use TextField?

Comment (datnguyen0126) [inline src/payments/serializers.py:23]: Why don't you use TextField?

File: src/payments/tests.py [hunk 3]
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
+
+# Create your tests here.
```
Comment (datnguyen0126) [inline src/payments/tests.py]: remove this

File: src/payments/services.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    def retrieve_payment_intent(cls, payment_intent_id):
+        if payment_intent_id is None:
+            raise Exception('payment_intent_id not found')
+        cancel_payment_intent = stripe.PaymentIntent.retrieve(
+            payment_intent_id,
+        )
+        if cancel_payment_intent is None:
+            raise Exception('retrieve payment intent wrongs')
+
+        return cancel_payment_intent
```
Comment (datnguyen0126) [inline src/payments/services.py:117]: new line

File: src/payments/views.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+            paginated_querySet = paginator.page(page)
+        except EmptyPage:
+            paginated_querySet = paginator.page(paginator.num_pages)
+        serializer = self.serializer_class(paginated_querySet, many=True)
+        content = {
+            "page": page,
+            "pagetotal": page_total,
+            "list_payments": serializer.data,
+        }
+        return Response(content)
```
Comment (datnguyen0126) [inline src/payments/views.py]: new line

**Review by NQLuan — APPROVED** [review]

**Review by datnguyen0126 — APPROVED** [review]

File: src/payments/migrations/0001_initial.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
+        ('tours', '0002_auto_20200805_1011'),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='Payment',
+            fields=[
+                ('id', models.UUIDField(default=uuid.uuid4, editable=False, primary_key=True, serialize=False, unique=True)),
+                ('payment_id', models.CharField(blank=True, max_length=255, null=True)),
```
Comment (prd-anh-nguyen) [inline src/payments/migrations/0001_initial.py:23]: why in the Payment table have both id and payment_id?

File: src/payments/services.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+            email=user.email,
+            name=data.get('name'),
+            phone=user.phone,
+        )
+        if customer is None:
+            raise Exception('Create customer wrongs')
+        user.stripe_id = customer.stripe_id
+        user.save()
+        del customer.id
+        del user.stripe_id
```
Comment (prd-anh-nguyen) [inline src/payments/services.py]: why do you delete `customer.id` and `user.stripe_id`

File: src/payments/views.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+                return Response(data=data, status=status.HTTP_201_CREATED)
+            except Exception as e:
+                return Response({'error_msg': str(e)}, status=status.HTTP_400_BAD_REQUEST)
+
+
+class PaymentIntent(APIView):
+
+    def post(self, request):
+        tour_id = request.query_params.get('tour_id')
+        tour = Tours.objects.get(id=tour_id)
```
Comment (prd-anh-nguyen) [inline src/payments/views.py]: It `Tours.objects.get(id=tour_id)` raise exception when have not found any tour object?

File: src/payments/views.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+    def get(self, request, format=None):
+        payments = Payments.objects.all().order_by('created_at')
+        pagesize = int(settings.PAGESIZE)
+        page_total = round(Payments.objects.all().count() / pagesize + 0.5)
+        paginator = Paginator(payments, pagesize)
+        page = request.GET.get("page", 1)
+        try:
+            paginated_querySet = paginator.page(page)
+        except PageNotAnInteger:
```
Comment (prd-anh-nguyen) [inline src/payments/views.py]: have you any idea to remove this exception `except PageNotAnInteger`?

File: src/payments/views.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+class PaymentsList(APIView):
+    serializer_class = PaymentSerializer
+    filter_backends = [filters.SearchFilter]
+    search_fields = ['title', 'address']
+
+    def get(self, request, format=None):
+        payments = Payments.objects.all().order_by('created_at')
+        pagesize = int(settings.PAGESIZE)
+        page_total = round(Payments.objects.all().count() / pagesize + 0.5)
```
Comment (prd-anh-nguyen) [inline src/payments/views.py]: I think paginator will support it, please research it

File: src/payments/services.py (hunk 7 above)
Comment (ptran068) [reply src/payments/services.py]: i think , it's important, and  make user account safully

File: src/payments/views.py (hunk 8 above)
Comment (ptran068) [reply src/payments/views.py]: i checked it at line 83

File: src/payments/views.py (hunk 10 above)
Comment (ptran068) [reply src/payments/views.py]: kkk ok , i will make it better

**Review by prd-anh-nguyen — APPROVED** [review]

**Pull request merged**